### PR TITLE
Move LED define to global + defines to fix TTGO compile

### LINKED ISSUE #37: Compilation error for TTGO T7 V1.3
image

TTGO T7 V1.3 framework defines LED_BUILTIN const instead of definition, this causes the conflict with LED_BUILTIN definition in consts.h in SlimeVR firmware. We may want to add this board type specifically and create separate set of definitions for it.

### PR DESCRIPTION
**Pull request by unlogisch04**
the LED_PIN is wrong in const.h
moved it to globals.h with the option to accept changes from defines.h
with that it should fix #37 

my understanding is, that globals.h adds all the missing stuff from defines.h with the default values.
Any meaning on this?

### CODE REVIEW

File: src/globals.h [hunk 1]
```diff
@@ -40,7 +40,27 @@
 #define BATTERY_MONITOR BAT_INTERNAL
 #endif
 
-#if defined(LED_INVERTED) && LED_INVERTED
+// If LED_PIN is not defined in "defines.h" take the default pin from "pins_arduino.h" framework.
+// If there is no pin defined for the board, use LED_PIN 255 and disable LED
+#if !defined(LED_PIN)
+#if defined(LED_BUILTIN)
```
Comment (TheDevMinerTV) [inline src/globals.h:46]:
```suggestion
    #if defined(LED_BUILTIN)
```

File: src/globals.h [hunk 2]
```diff
@@ -40,7 +40,27 @@
...
+#define LED_PIN LED_BUILTIN
```
Comment (TheDevMinerTV) [inline src/globals.h:47]:
```suggestion
        #define LED_PIN LED_BUILTIN
```

**Comment by Eirenliel** [comment]
It's not wrong for ESP-12f: https://randomnerdtutorials.com/esp8266-pinout-reference-gpios/

File: src/globals.h [hunk 3]
```diff
@@ -40,7 +40,27 @@
...
+#define ENABLE_LEDS true
```
Comment (TheDevMinerTV) [inline src/globals.h:48]:
```suggestion
        #define ENABLE_LEDS true
```

Also, what if `ENABLE_LEDS` is set to `false` in `defines.h`?

File: src/globals.h [hunk 4]
```diff
@@ -40,7 +40,27 @@
...
+#else
+#define LED_PIN 255
+#define ENABLE_LEDS false
+#endif
+#else
+// LED_PIN is defined
+#define ENABLE_LEDS true
```
Comment (TheDevMinerTV) [inline src/globals.h:55]:
```suggestion
    #define ENABLE_LEDS true
```

File: src/globals.h [hunk 5]
```diff
@@ -40,7 +40,27 @@
...
+#if !defined(LED_PIN)
+#if defined(LED_BUILTIN)
+#define LED_PIN LED_BUILTIN
+#define ENABLE_LEDS true
+#else
+#define LED_PIN 255
+#define ENABLE_LEDS false
+#endif
+#else
+// LED_PIN is defined
```
Comment (TheDevMinerTV) [inline src/globals.h:54]:
```suggestion
    // LED_PIN is defined
```

File: src/globals.h [hunk 6]
```diff
@@ -40,7 +40,27 @@
...
 #endif
 
-#if defined(LED_INVERTED) && LED_INVERTED
+// If LED_PIN is not defined in "defines.h" take the default pin from "pins_arduino.h" framework.
+// If there is no pin defined for the board, use LED_PIN 255 and disable LED
+#if !defined(LED_PIN)
+#if defined(LED_BUILTIN)
+#define LED_PIN LED_BUILTIN
+#define ENABLE_LEDS true
+#else
```
Comment (TheDevMinerTV) [inline src/globals.h:49]:
```suggestion
    #else
```

File: src/globals.h [hunk 7]
```diff
@@ -40,7 +40,27 @@
...
+#define LED_PIN 255
```
Comment (TheDevMinerTV) [inline src/globals.h:50]:
```suggestion
        #define LED_PIN 255
```

File: src/globals.h [hunk 8]
```diff
@@ -40,7 +40,27 @@
...
+#define ENABLE_LEDS false
+#endif
```
Comment (TheDevMinerTV) [inline src/globals.h:52]:
```suggestion
    #endif
```

File: src/globals.h [hunk 9]
```diff
@@ -40,7 +40,27 @@
...
-#if defined(LED_INVERTED) && LED_INVERTED
+// If LED_PIN is not defined in "defines.h" take the default pin from "pins_arduino.h" framework.
+// If there is no pin defined for the board, use LED_PIN 255 and disable LED
+#if !defined(LED_PIN)
+#if defined(LED_BUILTIN)
+#define LED_PIN LED_BUILTIN
+#define ENABLE_LEDS true
+#else
+#define LED_PIN 255
+#define ENABLE_LEDS false
```
Comment (TheDevMinerTV) [inline src/globals.h:51]:
```suggestion
        #define ENABLE_LEDS false
```

File: src/globals.h [hunk 10]
```diff
@@ -40,7 +40,27 @@
...
+#endif
+#else
+// LED_PIN is defined
+#define ENABLE_LEDS true
+#endif
+
+#if !defined(LED_INVERTED)
+// default is inverted for SlimeVR / ESP-12F
+#define LED_INVERTED true
```
Comment (TheDevMinerTV) [inline src/globals.h:60]:
```suggestion
    #define LED_INVERTED true
```

File: src/globals.h [hunk 11]
```diff
@@ -40,7 +40,27 @@
...
+#define LED_PIN 255
+#define ENABLE_LEDS false
+#endif
+#else
+// LED_PIN is defined
+#define ENABLE_LEDS true
+#endif
+
+#if !defined(LED_INVERTED)
+// default is inverted for SlimeVR / ESP-12F
```
Comment (TheDevMinerTV) [inline src/globals.h:59]:
```suggestion
    // default is inverted for SlimeVR / ESP-12F
```

**Review by TheDevMinerTV — CHANGES_REQUESTED** [review]

**Comment by Eirenliel** [comment]
These suggestings aren't helpful... Why do we need all of this? :nya_a: They probably should be part of the board definition idk. LED_BUILTIN should be defined for ESP from the toolset if i'm right anyway?

**Comment by TheDevMinerTV** [comment]
The TTGO things implement `LED_BUILTINT` as `static constexpr LED_BUILTIN` which collides with on of our defines so that the code basically doesn't compile for the TTGO boards, the ESPs on the other hand just use a `#define` for this.

**Comment by Eirenliel** [comment]
I see, I hate it. That's still should be part of board definition, no? Also I think that's not TTGO, that's ESP32 behaviour?

**Comment by unlogisch04** [comment]
I want the most flexibility for the defines.h
so everyone can override the default pins in defines.h or not if he wants to stay with the board (Arduino framework definition)
I did try to stay compatible with the generator, so it works as it would work with the old defines.h

I mean not the number or value of it. For esp-12f that value is already defined in the Arduino framework.
It is just in the wrong spot in the config .

1. const.h gets included as fist.
2. (arduino_board_default)
3. defines.h 
4. globals.h
globals.h is referenced as endpoint for the values. (the defines generated from here will be used in the code)

**Comment by unlogisch04** [comment]
Updated: Now it can be part of the defines.h #define board config

File: src/globals.h [hunk 12]
```diff
@@ -40,7 +40,27 @@
 #define BATTERY_MONITOR BAT_INTERNAL
 #endif
 
-#if defined(LED_INVERTED) && LED_INVERTED
+// If LED_PIN is not defined in "defines.h" take the default pin from "pins_arduino.h" framework.
+// If there is no pin defined for the board, use LED_PIN 255 and disable LED
+#if !defined(LED_PIN)
+    #if defined(LED_BUILTIN)
+        #define LED_PIN LED_BUILTIN
+        #define ENABLE_LEDS true
```
Comment (TheDevMinerTV) [inline src/globals.h:48]: What if `ENABLE_LEDS` is set to `false`?

Comment (unlogisch04) [reply src/globals.h:48]: Good catch. I think i have to redo that whole part, it it should not get more complex.

Comment (unlogisch04) [reply src/globals.h:48]: I changed some stuff.  Can you review again?
I added for every board. Ignoring ENABLE_LEDS  (overwriting it)
  #define LED_PIN LED_BUILTIN
  #define LED_INVERTED true

**Review by TheDevMinerTV — CHANGES_REQUESTED** [review]

File: src/defines.h [hunk 13]
```diff
@@ -47,38 +47,52 @@
 // #define BATTERY_SHIELD_R1 100 // Board voltage divider resistor Ain to GND in kOhm
 // #define BATTERY_SHIELD_R2 220 // Board voltage divider resistor Ain to INPUT_BOARD in kOhm
 
+// LED configuration; uncomment if you need to change it
```
[the remark this answers is not in the document]
Comment (unlogisch04) [reply src/defines.h:50]: Valid. Would that be better 
```
// LED configuration:
//   LED_PIN
//     - Number or Symbol (D1,..) of the Output
//     - To turn off the LED, set LED_PIN to LED_OFF
//   LED_INVERTED 
//     - false for output 3.3V on high
//     - true for pull down to GND on high
```

File: src/globals.h [hunk 14]
```diff
@@ -40,7 +40,31 @@
...
+    #else
+        #define LED_PIN 255
```
Comment (deiteris) [inline src/globals.h:50]:
```suggestion
        #define LED_PIN LED_OFF
```

File: src/defines.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (deiteris) [reply src/defines.h:50]: Yeah, the instructions would be better in this case.

File: src/globals.h [hunk 15]
```diff
@@ -40,7 +40,31 @@
...
+        #define ENABLE_LEDS false
+    #endif
+#else
+    // LED_PIN is defined
+    #if (LED_PIN < 0) || (LED_PIN >= 255)
```
Comment (TheDevMinerTV) [inline src/globals.h:55]:
```suggestion
    #if (LED_PIN < 0) || (LED_PIN >= LED_OFF)
```

**Review by TheDevMinerTV — CHANGES_REQUESTED** [review]

File: src/globals.h [hunk 16]
```diff
@@ -40,7 +40,31 @@
...
+    // LED_PIN is defined
+    #if (LED_PIN < 0) || (LED_PIN >= LED_OFF)
+        #define ENABLE_LEDS false    
+    #else
+        #define ENABLE_LEDS true
+    #endif
+#endif
+
+#if !defined(LED_INVERTED)
+    // default is inverted for SlimeVR / ESP-12F
```
Comment (TheDevMinerTV) [inline src/globals.h:63]:
```suggestion
    // default is inverted for SlimeVR / ESP-12E
```

File: src/globals.h [hunk 17]
```diff
@@ -40,7 +40,31 @@
 #define BATTERY_MONITOR BAT_INTERNAL
 #endif
 
-#if defined(LED_INVERTED) && LED_INVERTED
+// If LED_PIN is not defined in "defines.h" take the default pin from "pins_arduino.h" framework.
+// If there is no pin defined for the board, use LED_PIN 255 and disable LED
+#if !defined(LED_PIN)
```
Comment (TheDevMinerTV) [inline src/globals.h:45]: We should probably invert these branches, just to make it look a bit neater.

**Review by TheDevMinerTV — CHANGES_REQUESTED** [review]

File: src/globals.h (hunk 17 above)
Comment (unlogisch04) [reply src/globals.h]: ok done

**Review by TheDevMinerTV — APPROVED** [review]

**Comment by Eirenliel** [comment]
wow what happened here... good job y'all :D

**Pull request merged**